### PR TITLE
Withdraw all functionality and fail withdraw if not sufficient balance

### PR DESCRIPTION
**Pull request by bflood366**

### CODE REVIEW

File: backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt [hunk 1]
```diff
@@ -370,17 +378,37 @@ class BlockchainTransactionHandler(
...
+    private fun completeWithdrawals(settlingWithdrawals: List<WithdrawalEntity>, withdrawalEvents: Map<Int, Exchange.WithdrawalEventResponse> = emptyMap()) {
         settlingWithdrawals.forEach {
             try {
-                onTxComplete(it.transactionData!!, error)
+                val withdrawAmount = withdrawalEvents[it.sequenceId]?.let { event ->
+                    if (it.amount == BigInteger.ZERO && it.actualAmount != null && event.amount.compareTo(it.actualAmount) != 0) {
+                        logger.warn { "Withdraw all mismatch, sequencer=${it.actualAmount} contract=${event.amount} " }
+                    }
+                    event.amount
+                } ?: it.actualAmount ?: it.amount
```
Comment (bholzman) [inline backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt:396]: under what conditions would there not be an event? would it be better to log an error and not complete the withdrawal in that case?

File: backend/src/main/kotlin/co/chainring/core/db/Migrations.kt [hunk 2]
```diff
@@ -76,4 +77,5 @@ val migrations = listOf(
     V35_RemoveExchangeContract(),
     V36_AddPendingDepositsToTelegramBotUser(),
     V37_AddBatchHash(),
+    V38_AddActualAmountToWithdrawalTable(),
```
Comment (bholzman) [inline backend/src/main/kotlin/co/chainring/core/db/Migrations.kt:80]: fyi I just merged a change that added a V38_ migration

**Review by bholzman — APPROVED** [review]
Looks great!

File: backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt (hunk 1 above)
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt:396]: yeah there should always be an event. I will adjust this - it was primarily defensive, but better to fail fast if this happens and alert.

File: backend/src/main/kotlin/co/chainring/core/db/Migrations.kt (hunk 2 above)
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/db/Migrations.kt:80]: thanks will adjust - also there is a contract PR with this - so can't merge til that is reviewed - https://github.com/Chainring-Inc/chainring-contracts/pull/15

File: backend/src/main/kotlin/co/chainring/core/evm/EIP712Transaction.kt [hunk 3]
```diff
@@ -69,7 +70,13 @@ sealed class EIP712Transaction {
 
         override fun getTxData(sequence: Long): ByteArray {
             return serializeTx(
-                ExchangeTransactions.WithdrawWithSignature(ExchangeTransactions.Withdraw(sequence, sender.value, token.address.value, amount, nonce.toBigInteger()), signature.toByteArray()),
+                ExchangeTransactions.WithdrawWithSignature(ExchangeTransactions.Withdraw(sequence, sender.value, token.address.value, amount, nonce.toBigInteger()), signature.toByteArray(), BigInteger.ZERO),
+            )
+        }
+
+        fun getTxData(sequence: Long, withdrawAllAmount: BigInteger = BigInteger.ZERO): ByteArray {
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/co/chainring/core/evm/EIP712Transaction.kt:77]: I don't quite understand why do we need `withdrawAllAmount` to be honest. It is rather confusing tbh to have 2 amount params in the `withdraw` function of the contract. I've left a similar comment in the contract PR.

Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/evm/EIP712Transaction.kt:77]: I commented in the contracts PR what the rationale was for this approach

File: backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt (hunk 1 above)
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt:396]: Actually there is a case where there might not an event albeit rare - if anvil restarts while a withdrawal batch is in progress we do not get a receipt and the code assumes all transactions in the batch were successful - for now I changed to log if we do not get an event, but don't stop or fail anything,

File: backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt [hunk 4]
```diff
@@ -226,28 +225,34 @@ class BlockchainTransactionHandler(
...
-                                settlingWithdrawals.firstOrNull { it.sequenceId.toBigInteger() == failedWithdrawal.sequence }?.let {
-                                    completeWithdrawals(listOf(it), errorMsg)
+                            }.toMap()
+                            val completedWithdrawalBySequence = txReceipt.logs.mapNotNull { eventLog ->
+                                Contract.staticExtractEventParameters(
+                                    Exchange.WITHDRAWAL_EVENT,
+                                    eventLog,
+                                )?.let {
+                                    val event = Exchange.getWithdrawalEventFromLog(eventLog)
+                                    Pair(event.sequence.toInt(), event)
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt:243]: changed to a long

File: backend/src/main/kotlin/co/chainring/core/evm/EIP712Transaction.kt (hunk 3 above)
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/evm/EIP712Transaction.kt:77]: I updated the contracts and adjusted here to not send a separate withdrawAllAmount value.

**Pull request merged**